src/upload-button.tsx

```diff
@@ -0,0 +1,377 @@
+    };
+
+    const onUpload = async (event: React.ChangeEvent<HTMLInputElement>) => {
+        cockpit.assert(event.target.files, "not an <input type='file'>?");
+        let next_progress = 0;
+        const toUploadFiles = [];
+        console.log(event.target.files);
+
+        const resetInput = () => {
+        // Reset input field in the case a download was cancelled and has to be re-uploaded
+        // https://stackoverflow.com/questions/26634616/filereader-upload-same-file-again-not-working
+            event.target.value = "";
+        };
+
+        // Before uploading, ask the user if any existing files should be overwritten or skipped.
+        // Inefficient....
+        let resolution;
+        for (let i = 0; i < event.target.files.length; i++) {
+            const inputFile = event.target.files[i];
+
+            if (resolution && resolution.applyToAll && resolution.skip)
+                break;
+
+            if (resolution && resolution.applyToAll && resolution.replace)
+                toUploadFiles.push(inputFile);
+            else {
+                for (const file of files) {
+                    if (file.name === inputFile.name) {
+                        const waiter = new Waiter();
+                        Dialogs.show(
+                            <FileConflictDialog
+                              path={path}
+                              file={file}
+                              uploadFile={inputFile}
+                              waiter={waiter}
+                              showApplyAll={event.target.files.length > 1}
+                            />
+                        );
+
+                        try {
+                            resolution = await waiter.block();
+                        } catch (exc) {
+                            resetInput();
+                            return;
+                        }
+
+                        break;
+                    }
+                }
+
+                if (resolution && resolution.skip)
+                    continue;
+
+                toUploadFiles.push(inputFile);
+            }
+        }
+
+        if (toUploadFiles.length === 0) {
+            resetInput();
+            return;
+        }
+
+        window.addEventListener("beforeunload", beforeUnloadHandler);
+        console.log("to upload", toUploadFiles);
+
+        const cancelledUploads = [];
+        await Promise.allSettled(toUploadFiles.map(async (file: File) => {
+            const tmp_path = path.slice();
+            tmp_path.push(file.name);
+            const destination = tmp_path.join('/');
+            const abort = new AbortController();
+
+            setUploadedFiles(oldFiles => {
+                return {
+                    [file.name]: { file, progress: 0, cancel: () => abort.abort() },
+                    ...oldFiles,
+                };
+            });
+
+            try {
+                await upload(destination, file.stream(), (progress) => {
```

Comment:
```suggestion
                await upload(destination, file, (progress) => {
```

This will have to change after https://github.com/cockpit-project/cockpit/pull/20394 lands.

Reply:
https://github.com/cockpit-project/cockpit-files/pull/415

Reply:
And another broken change is that progress reporting does not work at all anymore? So I have to wait another cycle on a Cockpit release to write tests.

Reply:
I really think c-files image prepare should pull in cockpit from https://copr.fedorainfracloud.org/coprs/g/cockpit/main-builds/

Reply:
Right, what I'll do is introduce TEST_SCENARIO=upload and then do:
```
Makefile:VM_CUSTOMIZE_FLAGS += --run-command 'dnf -y copr enable rhcontainerbot/podman-next >&2; dnf -y update --repo "copr*" >&2'
```

Like podman but then with Cockpit and trigger fedora-40/upload just to verify things in CI it should not be a lot of extra work.

Reply:
No, easier is `TEST_SCENARIO=pybridge make vm` :)